yang/goldstone-nexthop.yang

```diff
@@ -0,0 +1,166 @@
+    prefix inet;
+  }
+
+  organization
+    "GoldStone";
+
+  description
+    "Goldstone Nexthop";
+
+  revision 2022-12-14 {
+    description
+      "Initial revision.";
+  }
+
+  typedef optional-ip-address {
+    type union {
+      type inet:ipv4-address;
+      type string {
+        length "0";
+      }
+    }
+  }
+
+  typedef nexthop-type {
+    type enumeration {
+      enum "interface" {
+        description
+          "Specific interface.";
+      }
+      enum "ip4" {
+        description
+          "IPv4 address.";
+      }
+      enum "ip4-interface" {
+        description
+          "IPv4 address and interface.";
+      }
+      enum "blackhole" {
+        description
+          "Unreachable or prohibited.";
+      }
+    }
+    description
+      "Nexthop types.";
+  }
+
+  typedef blackhole-type {
+    type enumeration {
+      enum "unspec" {
+        description
+          "Generic unreachable.";
+      }
+      enum "null" {
+        description
+          "Null type.";
+      }
+      enum "reject" {
+        description
+          "ICMP unreachable.";
+      }
+      enum "prohibited" {
+        description
+          "ICMP admin-prohibited.";
+      }
+    }
+    default "null";
+    description
+      "Nexthop blackhole types.";
+  }
+
+  grouping nexthop-config {
+    leaf nexthop-type {
+      type nexthop-type;
+      mandatory true;
+      description
+        "The nexthop type.";
+    }
+
+    leaf gateway {
+      type optional-ip-address;
+      description
+        "The nexthop gateway address.";
+    }
+
+    leaf interface {
+      type gs-if:interface-ref;
+      description
+        "The nexthop egress interface.";
+    }
+
+    leaf blackhole-type {
+      type blackhole-type;
+      description
+        "A blackhole sub-type, if the nexthop is a blackhole type.";
+    }
+
+    leaf onlink {
+      type boolean;
+      default "false";
+      description
+        "Nexthop is directly connected.";
+    }
+  }
+
+  grouping nexthop-state {
+    leaf active {
+      type boolean;
+      description
+        "Nexthop is active.";
+    }
+
+    leaf fib {
+      type boolean;
+      description
+        "Nexthop is installed in fib.";
+    }
+  }
+
+  grouping nexthop-top {
+    container nexthops {
+      list nexthop {
+        key "nexthop-type gateway interface";
+        min-elements 1;
+        description
+          "A list of nexthop objects.";
+
+        leaf nexthop-type {
+          type leafref {
+            path "../config/nexthop-type";
+          }
+        }
+
+        leaf gateway {
+          type leafref {
+            path "../config/gateway";
+          }
+        }
+
+        leaf interface {
+          type leafref {
+            path "../config/interface";
+          }
+        }
+
+        container config {
+          uses nexthop-config;
+        }
+        container state {
+          config false;
+          uses nexthop-config;
+          uses nexthop-state;
```

Comment:
Maybe the operational data for static-route/nexthop would be better placed in a centralized RIB/FIB structure that consolidates all future protocols routing base.
Is there any plan in this direction?

Reply:
@ipi-claytonpascoal
I thought your question here is related to what I described [here](https://github.com/oopt-goldstone/goldstone-mgmt/pull/73#issue-1503828799).
> We may need to decide if we had better to add such structure in current phase (of targeting on static-route support as the first step).

I agree with your comment as OcNOS, FRR,  and OpenConfig adopt such style and structure.

@ishidawataru 
Do you have any advise or comments about this?

Reply:
@KazuyaAnazawa First, let's keep the following `config`/`state` structure (OpenConfig style) in all containers of the Goldstone primitive models.
Following this structure makes the model structure more predictable when reading.

```yang
grouping b-config {
}

grouping b-state {
// can be empty
}

grouping b-top {
  container b {
    container config {
      uses b-config;
    }
    container state {
      uses b-config;
      uses b-state;
    }
  }
}

grouping a-config {
}

grouping a-state {
// can be empty
}

container a {
  container config {
    uses a-config;
  }
  container state {
    config false;
    uses a-config;
    uses a-state;
  }

  uses b-top;
}
```

At this moment, I think it's OK to keep operational information on the static route and next-hop in the corresponding state containers.
We will need to define a model for a RIB (and maybe FIB (it looks like OpenConfig doesn't have a FIB model)) when we support dynamic routing protocols. When we do that, we can consider whether it's better to move operational info to the RIB structure.

Goldstone primitive YANG model is an internal API for Goldstone management components. We're free to change it when we need to.

Also, I think we will end up refining the models when we start implementing the south daemon that supports them.
Let's don't put too much time into just defining a model. We can discuss this again when we have an actual implementation.

Reply:
@ishidawataru Thank you so much for your valuable comments and advise.
I first try to define the models so that we can keep OpenConfig style, and submit the early version of the models.